src/manifold/deferred.clj

```diff
@@ -1386,6 +1512,152 @@
+  (assert-some operator)
+  (chain this #(success-deferred
+                (.apply operator %))))
+
+(def ^:private then-apply-async (async-for then-apply))
+
+
+(defn- then-accept [this ^Consumer operator]
+  (assert-some operator)
+  (chain this #(.accept operator %)))
+
+(def ^:private then-accept-async (async-for then-accept))
+
+
+(defn- then-run [this ^Runnable operator]
+  (assert-some operator)
+  (chain this (fn [_] (.run operator))))
+
+(def ^:private then-run-async (async-for then-run))
+
+
+(defn- then-combine [this other ^BiFunction operator]
+  (assert-some other operator)
+  (let-flow [mine this
+             theirs other]
+    (success-deferred (.apply operator mine theirs))))
+
+(def ^:private then-combine-async (async-for-dual then-combine))
+
+
+(defn- then-accept-both [this other ^BiConsumer operator]
+  (assert-some other operator)
+  (let-flow [mine this
+             theirs other]
+    (.accept operator mine theirs)))
+
+(def ^:private then-accept-both-async (async-for-dual then-accept-both))
+
+
+(defn- run-after-both [this other ^Runnable operator]
+  (assert-some other operator)
+  (let-flow [_ this
+             _ other]
+    (.run operator)))
+
+
+(def ^:private run-after-both-async (async-for-dual run-after-both))
+
+
+(defn- apply-to-either [this other ^java.util.function.Function operator]
+  (assert-some other operator)
+  (then-apply (alt this other) operator))
+
+(def ^:private apply-to-either-async (async-for-dual apply-to-either))
+
+
+(defn- accept-either [this other ^java.util.function.Function operator]
+  (assert-some other operator)
+  (then-accept (alt this other) operator))
+
+(def ^:private accept-either-async (async-for-dual accept-either))
+
+
+(defn- run-after-either [this other ^java.util.function.Function operator]
+  (assert-some other operator)
+  (then-run (alt this other) operator))
+
+(def ^:private run-after-either-async (async-for-dual run-after-either))
+
+
+(defn- then-compose [this ^java.util.function.Function operator]
+  (assert-some operator)
+  (chain this (fn [value] (.apply operator value))))
+
+(def ^:private then-compose-async (async-for then-compose))
+
+
+(defn- then-handle [this ^java.util.function.BiFunction operator]
+  (assert-some operator)
+  (-> this
+      (chain #(.apply operator % nil))
+      (catch #(.apply operator nil %))))
+
+(def ^:private then-handle-async (async-for then-handle))
+
+
+(defn- then-exceptionally [this ^java.util.function.Function operator]
+  (assert-some operator)
+  (catch this #(.apply operator %)))
+
+(defn- to-completable-future [this]
+
+  (let [result (java.util.concurrent.CompletableFuture.)]
+
+    (on-realized this
+                 #(.complete result %)
+                 #(.completeExceptionally result %))
+
+    result))
+
+(defn- when-complete [this ^java.util.function.BiConsumer operator]
+  (assert-some operator)
+  (on-realized this
+               #(.accept operator % nil)
+               #(.accept operator nil %))
+  this)
```

Comment:
This violates the contract of `whenComplete`:

> When this stage is complete, the given action is invoked with the result (or null if none) and the exception (or null if none) of this stage as arguments. The returned stage is completed when the action returns.
>
> Unlike method handle, this method is not designed to translate completion outcomes, so the supplied action should not throw an exception. However, if it does, the following rules apply: if this stage completed normally but the supplied action throws an exception, then the returned stage completes exceptionally with the supplied action's exception. Or, if this stage completed exceptionally and the supplied action throws an exception, then the returned stage completes exceptionally with this stage's exception.

1. It (usually) returns the same value/exception that was passed in, but it's not supposed to complete until the operator/action finishes. Returning `this` is incorrect.
2. If `this` succeeded, but `operator` throws an exception, the returned stage should complete exceptionally with the operator's exception, which it's not doing here. OTOH, if both `this` and the `operator` failed, then the returned stage's exception should be `this`'s, _not_ the `operator`'s.